### PR TITLE
feature: add fingerprint integration

### PR DESCRIPTION
**Pull request by Kudo**
# Why

to dogfood `@expo/fingerprint`, this pr tries to the fingerprint integration.
close [ENG-8599](https://linear.app/expo/issue/ENG-8599/integrate-expofingerprint-to-expo-github-action)

for updating doc, i will do it in a separated pr.

# How

this pr introduces two new actions, mainly designed for pull requests.

## fingerprint

the action will check the fingerprint from the pr head commit against base commit and just output the fingerprint diff. this could be used for something like [pr labeler](https://github.com/Kudo/eas-app/blob/main/.github/workflows/pr-labeler.yml):

- fingerprint not changed
  Screenshot 2023-09-14 at 7 19 25 PM

- fingerprint changed
  Screenshot 2023-09-14 at 7 19 31 PM

## preview-build

the action will check the fingerprint from the pr head commit against base commit and create eas builds if necessary. [example workflow](https://github.com/Kudo/eas-app/blob/main/.github/workflows/pr-preview-eas-build.yml)

- fingerprint not changed
  Screenshot 2023-09-14 at 7 24 42 PM

- fingerprint not changed and found compatible EAS Build
  Screenshot 2023-09-19 at 2 02 00 AM

- fingerprint changed
  Screenshot 2023-09-14 at 7 24 32 PM

## Implementations

check the high-level design first: https://www.notion.so/expo/expo-github-action-preview-build-92a444c9e4b84726a85b0bf4ef619f6c
to check whether the pr has fingerprint changes, we keep track `gitCommitSha <-> fingerprint <-> easBuildId`  in a sqlite database and keep it in the github actions cache. the remaining part are just github integration like commenting or eas build command executions.

some challenges i had in this pr:

### **ncc**

i want to import the @expo/fingerprint from the version user specified in yaml, not the version prebundled when we use ncc to build. this is more like a dynamic require for ncc. other than that, sqlite3 has some prebuilt native node-pre-gyp
binaries. if i trigger ncc build on macos, the darwin binary cannot run on github linux or windows runners. to do this, mainly have two points here:

- making external require from ncc
- we don't want to install sqlite3 or @expo/fingerprint into user's project root. rather than that we install to github tool directory. we should further mutate `process.env['NODE_PATH']` to let nodejs further search the github tool directory when using `require('sqlite3')`.

### **github actions cache mutation**

github actions cache is designed to be immutable. as long as the cache is changed, you should have a new cache key. i don't want to pollute user's cache and have many `fingerprint-db-v1`, `fingerprint-db-v2`, `fingerprint-db-v3` and so on. i just have to use the github api to remove the cache key before saving new database. the downside is that we need `actions: write` permission when integrating the action.

### **github actions cache matching scope**

from the design of github actions cache, when saving cache in a branch, they will actually [save the cache key in its branch's namespace](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache). when updating database in a pr branch, saving the fingerprint-db cache will not update the fingerprint-db cache on main branch. as the result, we have to add `on.push.branches: [main]` from the workflow and update fingerprint database when prs merged back and push to main branch. that make the main branch's database updated when new prs coming.

to further update `easBuildId` from pr branches in the database, when a pr merged back to main branch. we should use github api to find the associated pr from the merge commit, find the pr comment we added, and find the `easBuildId` in the hidden metadata in the comment.

## Risks

what if concurrent github actions workflows pushed back to main branch, the database may be overwritten in unexpected state. in this case, currently i have to use the [github actions concurrency](https://docs.github.com/en/actions/using-jobs/using-concurrency) to limit the concurrency group when pushing on main:

```
concurrency: fingerprint-${{ github.event_name != 'pull_request' && 'main' || github.run_id }}
```

# Test Plan

- adding some unit tests for db related code
- otherwise test the github integration on the example app: https://github.com/Kudo/eas-app
  - [workflows](https://github.com/Kudo/eas-app/tree/main/.github/workflows)
  - [pr that has fingerprint changes](https://github.com/Kudo/eas-app/pull/10)
  - [pr that is fingerprint compatible](https://github.com/Kudo/eas-app/pull/11)

### CODE REVIEW

File: ncc.js [hunk 1]
```diff
@@ -26,6 +26,7 @@ async function build() {
 
 async function compile(entry) {
   const { code, map, assets } = await ncc(entry, {
+    externals: ['@expo/fingerprint', 'module', 'sqlite3'],
```
Comment (byCedric) [inline ncc.js:29]: Let's move this only to apply when building these two actions. Right now, I think it also gets applied to the other actions, which may be unexpected.

One thing we can do is to create an externals dictionary, with the `entry` as key.

File: package.json [hunk 2]
```diff
@@ -27,16 +27,20 @@
     "@actions/github": "^5.1.1",
     "@actions/io": "^1.1.2",
     "@actions/tool-cache": "^2.0.1",
-    "semver": "^7.3.8"
+    "semver": "^7.3.8",
+    "sqlite3": "^5.1.6",
```
Comment (byCedric) [inline package.json:31]: Let me check if there is a reasonable way for us to limit this being installed in all workflows. Preferably, we isolate this as much as we can. But, if there is no way to do so, let's just keep it as is.

image

**Review by byCedric — COMMENTED** [review]
It looks good so far, but have some reservations about using SQLite through github caches 😄  We'll see how that goes in testing.

Can you add two readmes next to the `action.yml` that describes these actions? E.g. copy https://github.com/expo/expo-github-action/blob/main/preview/README.md and add a warning, [like this](https://github.com/expo/expo-github-action/tree/main/command#readme), that mentions it's still experimental?

File: package.json (hunk 2 above)
Comment (Kudo) [reply package.json]: good catch! i think we don't really need the sqlite3 here. replace it by `@types/sqlite3`. but we still have the fixed `sqlite3@^5.1.6` in the code. that's the version we want to install dynamically in github actions (with cache supported)

**Comment by Kudo** [comment]
> Can you add two readmes next to the `action.yml` that describes these actions? E.g. copy https://github.com/expo/expo-github-action/blob/main/preview/README.md and add a warning, [like this](https://github.com/expo/expo-github-action/tree/main/command#readme), that mentions it's still experimental?

sure thing! plan to add READMEs in a separated pr. will open the pr later.

File: package.json (hunk 2 above)
Comment (Kudo) [reply package.json]: oh no, it will break unit test. well, let me fallback to add sqlite3 to devDependencies.

**Comment by Kudo** [comment]
further added a feature to show a link to latest compatible eas build
Screenshot 2023-09-19 at 2 02 00 AM

**Review by byCedric — APPROVED** [review]

**Comment by yonitou** [comment]
Hi,
Sorry to comment on a closed PR but I'm trying to find a way to notify you guys : the `actions/cache` action needs to be updated because the action is not working anymore for 2 days now (=> https://github.com/actions/cache/discussions/1510)
I know I'm not the only one having the issue because I noticed that your workflow `PR labeler` on your own `expo/expo` repository is failing too in the `post-run` action : https://github.com/expo/expo/actions/runs/13327389659/job/37223563462 with the error below: 
```
Failed to save: Unable to reserve cache with key fingerprint-db, another job may be creating this cache. More details: Cache already exists. Scope: refs/heads/main, Key: fingerprint-db, Version: 30a065c722796dde573d8ae372ea626e255e02bafec65597fe70a1d3b666fbb0
```

**Comment by Kudo** [comment]
@yonitou thanks for the kind heads-up. didn't notice the ci problem until you share it. will follow up

**Comment by yonitou** [comment]
No problem, I opened a PR right here : https://github.com/expo/expo-github-action/pull/333
I tested it and it fixed the above issue but a new error appeared afterwards 
```
Failed to save: Failed to CreateCacheEntry: Received non-retryable error: Failed request: (409) Conflict: cache entry with the same key, version, and scope already exists
```
Pretty sure it's an issue from the `@actions/cache` repository related to this thread : https://github.com/actions/cache/issues/1541

**Comment by Kudo** [comment]
sorry i missed your pr until i created my pr. will add you as co-author from #334

**Pull request merged**